grr/server/grr_response_server/gui/ui/components/client/status_chip.ts

```diff
@@ -0,0 +1,44 @@
+import {Component, Input, ChangeDetectionStrategy, OnChanges, SimpleChanges} from '@angular/core';
+import {DateTime} from 'luxon';
+import {interval, Subject, merge} from 'rxjs';
+import {map} from 'rxjs/operators';
+
+/**
+ * Component displaying the status of a Client in a material chip.
+ */
+@Component({
+  selector: 'status-chip',
+  templateUrl: './status_chip.ng.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class StatusChip implements OnChanges {
+  private static readonly STATUS_OFFLINE = 'offline';
+  private static readonly STATUS_ONLINE = 'online';
+  private static readonly ONLINE_TRESHOLD_MINUTES = 15;
+
+  @Input() lastSeen?: Date;
+  private lastSeenChange$ = new Subject<void>();
+
+  // status observable that updates every second and when lastSeen changes
+  status$ = merge(interval(1000), this.lastSeenChange$).pipe(
+    map(_ => this.getStatus())
+  );
+
+  ngOnChanges(_changes: SimpleChanges): void {
+    this.lastSeenChange$.next()
+  }
+
+  getStatus(): string {
+    if (this.lastSeen === undefined) {
+      return StatusChip.STATUS_OFFLINE;
+    }
+
+    const timeDiff = DateTime.local()
```

Comment:
This code contains some [accidental complexity](https://en.wikipedia.org/wiki/No_Silver_Bullet). As a reader I wonder what `DateTime.local()` means (is it the same as now?). Also, why the need to specify `'minutes'` twice?

I suggest to use a [Duration](https://moment.github.io/luxon/docs/class/src/duration.js~Duration.html) object as `ONLINE_THRESHOLD` instead of an integer as `ONLINE_THRESHOLD_MINUTES`. This way, the unit is encoded in the threshold itself and not only in its variable name. Also, you can use `this.lastSeen.diffNow()` to calculate the elapsed time.

Reply:
That is a great idea! The second `'minutes'` was used because luxon has the ability to convert a `Duration` to the unit of choosing. Indeed, the first one was not actually necessary as I thought.